#Session #1 Homework

#Set up the environment
You need to import NumPy, Pandas, Matplotlib and Seaborn

In [ ]:
#libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#Question 1
What's the version of NumPy that you installed?

In [ ]:
np.__version__

'1.21.6'

#Question 2
What's the version of Pandas?

In [ ]:
pd.__version__

'1.3.5'

#Getting the data

Download it from [here](https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv).

In [ ]:
dataset = pd.read_csv('wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv')

#or
#!wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
#dataset = pd.read_csv('data.csv')

In [ ]:
dataset.head()

Make       Model  Year                Engine Fuel Type  Engine HP  \
0          BMW  1 Series M  2011     premium unleaded (required)      335.0   
1          BMW    1 Series  2011     premium unleaded (required)      300.0   
2          BMW    1 Series  2011     premium unleaded (required)      300.0   
3          BMW    1 Series  2011     premium unleaded (required)      230.0   
4          BMW    1 Series  2011     premium unleaded (required)      230.0   
...        ...         ...   ...                             ...        ...   
11909    Acura         ZDX  2012     premium unleaded (required)      300.0   
11910    Acura         ZDX  2012     premium unleaded (required)      300.0   
11911    Acura         ZDX  2012     premium unleaded (required)      300.0   
11912    Acura         ZDX  2013  premium unleaded (recommended)      300.0   
11913  Lincoln      Zephyr  2006                regular unleaded      221.0   

       Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
0                   6.0            MANUAL   rear wheel drive              2.0   
1                   6.0            MANUAL   rear wheel drive              2.0   
2                   6.0            MANUAL   rear wheel drive              2.0   
3                   6.0            MANUAL   rear wheel drive              2.0   
4                   6.0            MANUAL   rear wheel drive              2.0   
...                 ...               ...                ...              ...   
11909               6.0         AUTOMATIC    all wheel drive              4.0   
11910               6.0         AUTOMATIC    all wheel drive              4.0   
11911               6.0         AUTOMATIC    all wheel drive              4.0   
11912               6.0         AUTOMATIC    all wheel drive              4.0   
11913               6.0         AUTOMATIC  front wheel drive              4.0   

                             Market Category Vehicle Size  Vehicle Style  \
0      Factory Tuner,Luxury,High-Performance      Compact          Coupe   
1                         Luxury,Performance      Compact    Convertible   
2                    Luxury,High-Performance      Compact          Coupe   
3                         Luxury,Performance      Compact          Coupe   
4                                     Luxury      Compact    Convertible   
...                                      ...          ...            ...   
11909             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11910             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11911             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11912             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11913                                 Luxury      Midsize          Sedan   

       highway MPG  city mpg  Popularity   MSRP  
0               26        19        3916  46135  
1               28        19        3916  40650  
2               28        20        3916  36350  
3               28        18        3916  29450  
4               28        18        3916  34500  
...            ...       ...         ...    ...  
11909           23        16         204  46120  
11910           23        16         204  56670  
11911           23        16         204  50620  
11912           23        16         204  50920  
11913           26        17          61  28995  

[11914 rows x 16 columns]

#Question 3
What's the average price of BMW cars in the dataset?

In [ ]:
dataset[dataset['Make'] == 'BMW'].MSRP.describe()

count       334.000000
mean      61546.763473
std       27982.682153
min        4697.000000
25%       42300.000000
50%       51850.000000
75%       79725.000000
max      141200.000000
Name: MSRP, dtype: float64

mean row is the answer 61546.763473.

#Question 4
Select a subset of cars after year 2015 (inclusive, i.e. 2015 and after). How many of them have missing values for Engine HP?

In [ ]:
dataset[dataset['Year'] >= 2015].isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        0
Engine HP              51
Engine Cylinders        8
Transmission Type       0
Driven_Wheels           0
Number of Doors         5
Market Category      1324
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

There are 51 null values

#Question 5

* Calculate the average "Engine HP" in the dataset.
* Use the fillna method and to fill the missing values in "Engine HP" with the mean value from the previous step.
* Now, calcualte the average of "Engine HP" again.
* Has it changed?

Round both means before answering this questions. You can use the round function for that:

In [ ]:
#Calculate the average "Engine HP" in the dataset.
average = dataset['Engine HP'].mean()
average

249.38607007176023

In [ ]:
#Use the fillna method and to fill the missing values in "Engine HP" with the mean value from the previous step.
dataset['Engine HP'].fillna(value=average)

0        335.0
1        300.0
2        300.0
3        230.0
4        230.0
         ...  
11909    300.0
11910    300.0
11911    300.0
11912    300.0
11913    221.0
Name: Engine HP, Length: 11914, dtype: float64

In [ ]:
#Now, calcualte the average of "Engine HP" again.
average = dataset['Engine HP'].mean()
average

249.38607007176023

The average is not changed

#Question 6
* Select all the "Rolls-Royce" cars from the dataset.
* Select only columns "Engine HP", "Engine Cylinders", "highway MPG".
* Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 7 rows).
* Get the underlying NumPy array. Let's call it X.
* Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
* Invert XTX.
* What's the sum of all the elements of the result?

In [ ]:
#Select all the "Rolls-Royce" cars from the dataset.
dataset[dataset['Make'] == "Rolls-Royce"]

In [ ]:
#Select only columns "Engine HP", "Engine Cylinders", "highway MPG".
dataset[dataset['Make'] == "Rolls-Royce"][["Engine HP", "Engine Cylinders", "highway MPG"]]

In [ ]:
#Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 7 rows).
dataset[dataset['Make'] == "Rolls-Royce"][["Engine HP", "Engine Cylinders", "highway MPG"]].drop_duplicates()

Engine HP  Engine Cylinders  highway MPG
2921       325.0               8.0           15
3505       563.0              12.0           19
5275       563.0              12.0           21
5279       563.0              12.0           20
7443       322.0              12.0           15
7553       453.0              12.0           19
11448      624.0              12.0           21

In [ ]:
#Get the underlying NumPy array. Let's call it X
X = dataset[dataset['Make'] == "Rolls-Royce"][["Engine HP", "Engine Cylinders", "highway MPG"]].drop_duplicates().values
X

array([[325.,   8.,  15.],
       [563.,  12.,  19.],
       [563.,  12.,  21.],
       [563.,  12.,  20.],
       [322.,  12.,  15.],
       [453.,  12.,  19.],
       [624.,  12.,  21.]])

In [ ]:
#Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
XTX = X.T.dot(X)
XTX

array([[1.754801e+06, 3.965600e+04, 6.519600e+04],
       [3.965600e+04, 9.280000e+02, 1.500000e+03],
       [6.519600e+04, 1.500000e+03, 2.454000e+03]])

In [ ]:
#Invert XTX
XTX_inv = np.linalg.inv(XTX)

In [ ]:
#What's the sum of all the elements of the result?
XTX_inv.sum()

0.032212320677486125

#Questions 7
* Create an array y with values [1000, 1100, 900, 1200, 1000, 850, 1300].
* Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
* What's the value of the first element of w?.

In [ ]:
#Create an array  with values [1000, 1100, 900, 1200, 1000, 850, 1300]
y = [1000, 1100, 900, 1200, 1000, 850, 1300]

In [ ]:
#Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w
w = XTX_inv.dot(X.T).dot(y)

In [ ]:
#What's the value of the first element of w?
w[0]

0.19989598183188972